### PR TITLE
Fixes #5108 `DateValidator` resets `$timestampAttribute` value on empty attribute

### LINKED ISSUE #5108: DateValidator and timestampAttribute
`DateValidator` if attribute is empty `timestampAttribute` not changing. It would be good if it becomes null

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #5108

### CODE REVIEW

File: framework/validators/DateValidator.php [hunk 1]
```diff
@@ -266,6 +266,13 @@ public function init()
     public function validateAttribute($model, $attribute)
     {
         $value = $model->$attribute;
+        if ($this->isEmpty($value)) {
+            if ($this->timestampAttribute !== null) {
```
Comment (samdark) [inline framework/validators/DateValidator.php]: This check could be skipped. Assigning null is cheaper than checking.

**Review by samdark — APPROVED** [review]

File: framework/validators/DateValidator.php (hunk 1 above)
Comment (cebe) [inline framework/validators/DateValidator.php]: this is not the same, current code is correct and must not skip the check.

File: tests/framework/validators/DateValidatorTest.php [hunk 2]
```diff
@@ -200,10 +195,6 @@ public function testValidateAttributeICUFormat($timezone)
             $model->attr_timestamp
         );
         $val = new DateValidator(['format' => 'yyyy-MM-dd']);
-        $model = FakedValidationModel::createWithAttributes(['attr_date' => []]);
-        $val->validateAttribute($model, 'attr_date');
-        $this->assertTrue($model->hasErrors('attr_date'));
-        $val = new DateValidator(['format' => 'yyyy-MM-dd']);
```
Comment (cebe) [inline tests/framework/validators/DateValidatorTest.php]: why did you delete these tests?

File: framework/validators/DateValidator.php (hunk 1 above)
Comment (samdark) [inline framework/validators/DateValidator.php]: Oh. Right. I haven't noticed `{` :(

File: tests/framework/validators/DateValidatorTest.php (hunk 2 above)
Comment (klimov-paul) [inline tests/framework/validators/DateValidatorTest.php]: These are tests for empty value parsing failure, which contradicts the issue #5108.
This fix can not be applied keeping them.

Comment (cebe) [inline tests/framework/validators/DateValidatorTest.php]: I see what you mean now. If I recall correctly, these should cover robustness of the validator against array input.
The test should be kept for that reason and should be changed to test for non-empty array.

Comment (klimov-paul) [inline tests/framework/validators/DateValidatorTest.php]: Tests restored.

**Review by cebe — APPROVED** [review]

**Comment by cebe** [comment]
Thanks!
